compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -2567,27 +2567,30 @@ class Typer extends Namer
 
     def adaptNoArgs(wtp: Type): Tree = {
       val ptNorm = underlyingApplied(pt)
-      lazy val functionExpected = defn.isFunctionType(ptNorm)
-      lazy val resultMatch = constrainResult(tree.symbol, wtp, followAlias(pt))
+      def functionExpected = defn.isFunctionType(ptNorm)
+      def resultMatch = constrainResult(tree.symbol, wtp, followAlias(pt))
       def needsEta = pt match {
         case _: SingletonType => false
         case IgnoredProto(_: FunOrPolyProto) => false
         case _ => true
       }
+      var resMatch: Boolean = false
       wtp match {
         case wtp: ExprType =>
           readaptSimplified(tree.withType(wtp.resultType))
-        case wtp: MethodType if wtp.isImplicitMethod && (resultMatch || !functionExpected) =>
-          if (resultMatch || ctx.mode.is(Mode.ImplicitsEnabled)) adaptNoArgsImplicitMethod(wtp)
+        case wtp: MethodType if wtp.isImplicitMethod &&
+          ({ resMatch = resultMatch; resMatch } || !functionExpected) =>
+           if (resMatch || ctx.mode.is(Mode.ImplicitsEnabled)) adaptNoArgsImplicitMethod(wtp)
```

Comment:
oops I added an extra space in my initial suggestion:
```suggestion
          if (resMatch || ctx.mode.is(Mode.ImplicitsEnabled)) adaptNoArgsImplicitMethod(wtp)
```